update snapshot docs to include formal info, new bucket optimizations (#15134)
MystenLabs/sui: docs/content/guides/operator/snapshots.mdx

Snapshots of the Sui network enable Full node operators a way to bootstrap a Full node without having to execute all the transactions that occurred after genesis. You can upload snapshots to remote object stores like S3, Google Cloud Storage, Azure Blob Storage, and similar services. These services typically run the export process in the background so there is no degradation in performance for your Full node. With snapshots stored in the cloud, you're more easily able to recover quickly from catastrophic failures in your system or hardware.

To maintain a healthy Sui network, Sui encourages the Sui community to bring up additional snapshots to ensure stronger data availability across the network.

## Supported snapshot types

Sui supports two types of snapshots:

- **RocksDB snapshots** are a point-in-time view of a database store. This means that the snapshot keeps the state of the system at the moment it generates the snapshot, including non-pruned data, additional indices, and other data.

> - **Formal snapshots** are database agnostic: they keep the state of the system in the minimalistic format. The Formal snapshots also include cryptographic state accumulators, which are certified by the quorum of the validators. These can be verified by the node during the reconstruction. This is one of the main advantages over the RocksDB snapshots, which don't include any special integrity data certified by a quorum.

Reviewer: ```The Formal snapshots also include cryptographic state accumulators, which are certified by the quorum of the validators. These can be verified by the node during the reconstruction.```
Do we want to word it a little bit like - "The formal snapshots are verified during restore using cryptographic state accumulation by comparing against the state root hash from the verified end of epoch checkpoint summary". Also cc @williampsmith for suggestions if any

Reviewer: Absolutely, I just wanted us to have a conversation about what we all think should be in the docs or if it needs to be tweaked a little bit. Your suggestion sounds great!

To maintain a healthy Sui network, Sui encourages the Sui community to bring up additional snapshots to ensure stronger data availability across the network.

## Supported snapshot types

Sui supports two types of snapshots:

- **RocksDB snapshots** are a point-in-time view of a database store. This means that the snapshot keeps the state of the system at the moment it generates the snapshot, including non-pruned data, additional indices, and other data.
- **Formal snapshots** are database agnostic with a minimalistic format, i.e. they contain only the data necessary to restore a node to a valid state at the specified epoch, thus they have a much smaller storage footprint and are generally faster to restore from when compared to database snapshots. Formal snapshots are also supported natively by the Sui protocol. In this context, this means that we can cryptographically verify the contents of the formal snapshot against a commitment from the committee at the epoch we are restoring to. This verification happens automatically and by default during a formal snapshot restore (unless explicitly bypassed). 

> **Important Note:** formal snapshots are not suitable for use if you are running a RPC node that does any historical data lookups. For more information on this subject, see [node data management](./data-management.mdx).

Reviewer 2: ```suggestion
:::info

Formal snapshots are not suitable for use if you are running an RPC node that does any historical data lookups. For more information on node data management, see [Data Management](./data-management.mdx).

:::
```
Includes Will.i.am's suggestion.

- **RocksDB snapshots** are a point-in-time view of a database store. This means that the snapshot keeps the state of the system at the moment it generates the snapshot, including non-pruned data, additional indices, and other data.
- **Formal snapshots** are database agnostic with a minimalistic format, i.e. they contain only the data necessary to restore a node to a valid state at the specified epoch, thus they have a much smaller storage footprint and are generally faster to restore from when compared to database snapshots. Formal snapshots are also supported natively by the Sui protocol. In this context, this means that we can cryptographically verify the contents of the formal snapshot against a commitment from the committee at the epoch we are restoring to. This verification happens automatically and by default during a formal snapshot restore (unless explicitly bypassed). 

> **Important Note:** formal snapshots are not suitable for use if you are running a RPC node that does any historical data lookups. For more information on this subject, see [node data management](./data-management.mdx).

Reviewer 3: ```suggestion
**Important Note:** formal snapshots are not suitable for use if you are running an RPC node that does any historical data lookups. For more information on this subject, see [node data management](./data-management.mdx).
```

   - `bucket`: The S3 bucket name to store the snapshots.
   - `aws-access-key-id` and `aws-secret-access-key`: AWS authentication information with write access to the bucket.
   - `aws-region`: Region where bucket exists.
   - `object-store-connection-limit`: Number of simultaneous connections to the object store.

### Enabling Formal snapshots

Add an entry to the config file for `db-checkpoint-config`. Using Amazon's S3 service as an example:
   ```yaml
   state-snapshot-write-config:
      object-store-config:
         object-store: "S3"
         bucket: "<BUCKET-NAME>"
         aws-access-key-id: “<ACCESS-KEY>”
         aws-secret-access-key: “<SHARED-KEY>”
         aws-region: "<BUCKET-REGION>"
         object-store-connection-limit: 200
   ```
   The configuration settings show in the example are specific to AWS S3, but GCS, Azure Storage, and Cloudflare R2 are all supported.

## Restoring a Full node using snapshots

### Restoring using RocksDB snapshots

To restore from a RocksDB snapshot, follow these steps:

1. Download the snapshot for the epoch you want to restore to your local disk. There is one snapshot per epoch in s3 bucket.

> 1. Place the snapshot into the directory that the `db-config` value points to in your fullnode.yaml file. For example, if the `db-config` value points to `/opt/sui/db/authorities_db/full_node_db` and you want to restore from epoch 10, then copy the snapshot to the directory with this command:

Author: It can be whatever 🤷  . We use fullnode.yaml elsewhere in this guide so I changed them all to fullnode.yaml

> 1. Place the snapshot into the directory that the `db-config` value points to in your fullnode.yaml file. For example, if the `db-config` value points to `/opt/sui/db/authorities_db/full_node_db` and you want to restore from epoch 10, then copy the snapshot to the directory with this command:

Reviewer 3: The thing about using `fullnode.yaml` is that it sort of implies that this is only for fullnodes. But maybe it'll be more confusing to not be consistent.

:::

### Restoring using Formal snapshots

To restore using a Formal snapshot, use the `sui-tool` binary. `sui-tool` can be downloaded along with other sui binaries.
See [Install Sui](../developer/getting-started/sui-install.mdx) for more details.

The following steps can be used to restore a node from a formal snapshot:

1. If it's running, stop the node.
2. Run the command:
   ```

>    sui-tool download-db-snapshot --epoch "<EPOCH-NUMBER>" --genesis "<PATH-TO-GENESIS-BLOB>" \

Reviewer 3: Sorry to do this, but after our debugging with @tharbert it seems we will need to add `--network` in, or specify the archival bucket and bucket type as well (if they pass in `--network` it will default to using our archival bucket. (Not sure if it will be requester pays or not for archival - you'd probably know better)

Author: > Not sure if it will be requester pays or not for archival

No, not for the time-being

        --formal --network "<NETWORK>" --path <PATH-TO-NODE-DB> --num-parallel-downloads 50 --no-sign-request --snapshot-bucket-type=s3

>    ```

Reviewer 3: This command only showcases a way for them to use our hosted formal snapshots. But hopefully many will be storing and using their own. Could we start with showing the general case, and then after, show the special case (using ours), with a note that this would be using ours?

In the general case, we would omit `--network` as this allows `sui-tool` to generate sane defaults for ML buckets and such, and then we would add `--snapshot-bucket <bucket name>` and `--snapshot-bucket-type <type, e.g. gcs>`

   - `--epoch`: The epoch that you want to download, eg. `200`. Mysten Labs hosted buckets will only keep the last 90 epochs. 
   - `--genesis`: The path to the location of the network's `genesis.blob`.
   - `--formal`: If set, restore from formal (slim, DB agnostic) snapshot.
   - `--network`: Network to download snapshot for. Defaults to "mainnet".
   - `--path`: Path to snapshot directory on local filesystem.
   - `--no-sign-request`: If true, no authentication is needed for snapshot restores.
   - `--snapshot-bucket-type`: Snapshot bucket type. GCS and S3 currently supported.

## Mysten Labs managed snapshots

> Mysten is providing two tiers of snapshot storage access. *High throughput, Requester Pays enabled buckets*, and *Free, permissionless buckets*.

Reviewer 3: ```suggestion
Mysten Labs hosts two tiers of snapshot storage access. *High throughput, Requester Pays enabled buckets*, and *free, permissionless buckets*.
```
